xonsh/commands_cache.py

```diff
@@ -76,59 +89,97 @@ def remove_dups(p):
                 ret.append(e)
         return ret
 
-    @property
-    def all_commands(self):
-        paths = builtins.__xonsh__.env.get("PATH", [])
-        paths = CommandsCache.remove_dups(paths)
-        path_immut = tuple(x for x in paths if os.path.isdir(x))
+    def _update_if_changed(self, paths: tp.Tuple[str, ...], aliases):
         # did PATH change?
-        path_hash = hash(path_immut)
-        cache_valid_path = path_hash == self._path_checksum
+        path_hash = hash(paths)
+        yield path_hash == self._path_checksum
         self._path_checksum = path_hash
+
         # did aliases change?
-        alss = getattr(builtins, "aliases", dict())
-        al_hash = hash(frozenset(alss))
-        cache_valid_aliases = al_hash == self._alias_checksum
+        al_hash = hash(frozenset(aliases))
+        yield al_hash == self._alias_checksum
         self._alias_checksum = al_hash
+
         # did the contents of any directory in PATH change?
-        max_mtime = 0
-        for path in path_immut:
-            mtime = os.stat(path).st_mtime
-            if mtime > max_mtime:
-                max_mtime = mtime
-        cache_valid_paths = max_mtime <= self._path_mtime
+        max_mtime = max(map(lambda path: os.stat(path).st_mtime, paths), default=0)
+        yield max_mtime <= self._path_mtime
         self._path_mtime = max_mtime
 
-        if cache_valid_path and cache_valid_paths:
-            if not cache_valid_aliases:
+    @property
+    def all_commands(self):
+        paths = CommandsCache.remove_dups(builtins.__xonsh__.env.get("PATH", []))
+        path_immut = tuple(x for x in paths if os.path.isdir(x))
+        alss = getattr(builtins, "aliases", dict())
```

Comment:
this is inside a property. It is better to pass things, rather than depending on a global state. I know it is ugly to mix both. But here it is only for the property all_commands